### PR TITLE
Use type specific entity retrieval to replace predicates

### PR DESCRIPTION
**Pull request by 2No2Name**
In some places predicates are used to filter entities by type (armorstand cramming with minecarts, itemframes/paintings colliding with each other, llamas trying to find other llamas for caravans). With many entities of a different type nearby, evaluating the predicate on them wastes a lot of CPU time. Vanilla already provides a way to get only entities of a certain class, and making use of this method (adapted to allow excluding one entity) improves the performance a lot in certain configurations.

In my test with a 101x101 grid of armorstands standing on the ground, this optimization brought the 186mspt down to 145mspt. Of course that test is not a real gameplay situation, but it shows that getting entities of the wrong type and evaluating the predicate costs a lot of performance.

### CODE REVIEW

File: src/main/java/me/jellysquid/mods/lithium/common/world/WorldHelper.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        int i = MathHelper.floor((box.x1 - 2.0D) / 16.0D);
+        int j = MathHelper.ceil((box.x2 + 2.0D) / 16.0D);
+        int k = MathHelper.floor((box.z1 - 2.0D) / 16.0D);
+        int l = MathHelper.ceil((box.z2 + 2.0D) / 16.0D);
+        List<Entity> list = Lists.newArrayList();
+        ChunkManager chunkManager = world.getChunkManager();
+
+        for(int m = i; m < j; ++m) {
+            for(int n = k; n < l; ++n) {
+                WorldChunk worldChunk = chunkManager.getWorldChunk(m, n, false);
```
Comment (jellysquid3) [inline src/main/java/me/jellysquid/mods/lithium/common/world/WorldHelper.java]: These names should be de-obfuscated.

File: src/main/java/me/jellysquid/mods/lithium/common/world/WorldHelper.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+     *  [VanillaCopy] Method for getting entities by class but also exclude one entity
+     */
+    private static void getEntitiesOfClass(WorldChunk worldChunk, Entity excluded, Class<? extends Entity> entityClass, Box box, List<Entity> result) {
+        TypeFilterableList<Entity>[] entitySections = worldChunk.getEntitySectionArray();
+        int i = MathHelper.floor((box.y1 - 2.0D) / 16.0D);
+        int j = MathHelper.floor((box.y2 + 2.0D) / 16.0D);
+        i = MathHelper.clamp(i, 0, entitySections.length - 1);
+        j = MathHelper.clamp(j, 0, entitySections.length - 1);
+
+        for(int k = i; k <= j; ++k) {
```
Comment (jellysquid3) [inline src/main/java/me/jellysquid/mods/lithium/common/world/WorldHelper.java]: More obfuscated names.

**Review by jellysquid3 — CHANGES_REQUESTED** [review]

File: src/main/java/me/jellysquid/mods/lithium/common/world/WorldHelper.java (hunk 1 above)
Comment (2No2Name) [reply src/main/java/me/jellysquid/mods/lithium/common/world/WorldHelper.java]: I deobfuscated them now

**Review by jellysquid3 — APPROVED** [review]

**Comment by 2No2Name** [comment]
Merged after rebasing onto dev.

**Pull request closed without merging**